File: tests/Scribe.Core.Tests/DashNormalizerTests.cs
using Scribe.Core.Cleanup;

namespace Scribe.Core.Tests;

/// <summary>
/// The writing style asks the model not to emit em/en dashes, but that is advisory. These lock in the
/// deterministic backstop that makes the house style hold no matter which model answered.
/// </summary>
public class DashNormalizerTests
{
    private const string Em = "\u2014";
    private const string En = "\u2013";

    [Fact]
    public void Leaves_text_without_dashes_untouched()
    {
        const string input = "A normal sentence, with a comma; and a semicolon.";
        Assert.Same(input, DashNormalizer.Normalize(input));
    }

    [Fact]
    public void Hyphens_and_product_names_survive()
    {
        const string input = "Use GPT-5.6-Terra with Qwen3-14B on a well-structured run-on sentence.";
        Assert.Equal(input, DashNormalizer.Normalize(input));
    }

    [Fact]
    public void Spaced_em_dash_becomes_a_comma()
    {
        Assert.Equal(
            "The pill shows Scribe is listening, and where it lives.",
            DashNormalizer.Normalize($"The pill shows Scribe is listening {Em} and where it lives."));
    }

    [Fact]
    public void Unspaced_em_dash_becomes_a_comma_and_a_space()
    {
        Assert.Equal(
            "It works, mostly.",
            DashNormalizer.Normalize($"It works{Em}mostly."));
    }

    [Fact]
    public void Paired_em_dashes_both_become_commas()
    {
        Assert.Equal(
            "Say a phrase, like this one, and Scribe types it.",
            DashNormalizer.Normalize($"Say a phrase {Em} like this one {Em} and Scribe types it."));
    }

    [Fact]
    public void En_dash_between_numbers_becomes_a_range_word()
    {
        Assert.Equal("about 1 to 2 GB", DashNormalizer.Normalize($"about 1{En}2 GB"));
        Assert.Equal("pages 3 to 7", DashNormalizer.Normalize($"pages 3 {En} 7"));
    }

    [Fact]
    public void Does_not_double_up_existing_punctuation()
    {
        Assert.Equal("Wait, I mean the park.", DashNormalizer.Normalize($"Wait, {Em} I mean the park."));
        Assert.Equal("Done. Next up.", DashNormalizer.Normalize($"Done. {Em} Next up."));
    }

    [Fact]
    public void Line_leading_dash_is_dropped_rather_than_turned_into_a_comma()
    {
        Assert.Equal(
            "Items:\nfirst\nsecond",
            DashNormalizer.Normalize($"Items:\n{Em} first\n{Em} second"));
    }

    [Fact]
    public void Trailing_dash_is_removed_without_adding_punctuation()
    {
        Assert.Equal("An unfinished thought", DashNormalizer.Normalize($"An unfinished thought {Em}"));
    }

    [Fact]
    public void Dash_before_a_closing_bracket_or_quote_is_dropped_not_comma_ed()
    {
        Assert.Equal("(an aside)", DashNormalizer.Normalize($"(an aside {Em})"));
        Assert.Equal("he said \"fine\"", DashNormalizer.Normalize($"he said \"fine{Em}\""));
        Assert.Equal("[note]", DashNormalizer.Normalize($"[note{Em}]"));
    }

    [Fact]
    public void Dash_at_end_of_a_line_does_not_leave_a_dangling_comma()
    {
        Assert.Equal("First thought\nSecond thought", DashNormalizer.Normalize($"First thought {Em}\nSecond thought"));
        Assert.Equal("First\r\nSecond", DashNormalizer.Normalize($"First{Em}\r\nSecond"));
    }

    [Fact]
    public void Newlines_are_preserved()
    {
        Assert.Equal(
            "First para, with an aside.\n\nSecond para.",
            DashNormalizer.Normalize($"First para {Em} with an aside.\n\nSecond para."));
    }

    [Fact]
    public void Runs_of_dashes_collapse_to_one_replacement()
    {
        Assert.Equal("Yes, really.", DashNormalizer.Normalize($"Yes {Em}{Em} really."));
        Assert.Equal("Yes, really.", DashNormalizer.Normalize($"Yes {Em}{En} really."));
    }

    [Theory]
    [InlineData("")]
    [InlineData(null)]
    public void Null_and_empty_are_safe(string? input) =>
        Assert.Equal(string.Empty, DashNormalizer.Normalize(input));

    [Fact]
    public void Output_never_contains_a_dash()
    {
        string[] cases =
        [
            $"a {Em} b",
            $"a{Em}b",
            $"{Em}leading",
            $"trailing{Em}",
            $"1{En}2",
            $"one {Em} two {Em} three {Em} four",
            $"{Em}",
            $" {Em} ",
            $"(aside {Em})",
            $"line {Em}\nnext",
            $"quote{Em}\"",
        ];

        foreach (var c in cases)
        {
            Assert.False(DashNormalizer.ContainsDash(DashNormalizer.Normalize(c)), c);
        }
    }

    [Fact]
    public void Sanitized_cleanup_output_is_dash_free()
    {
        // The real guarantee: whatever the model answered, what reaches the document has no dashes.
        Assert.True(TextCleanupService.TrySanitize(
            $"The defaults are right for most people {Em} changes take effect after restart.",
            "the defaults are right for most people changes take effect after restart",
            out var text));

        Assert.Equal("The defaults are right for most people, changes take effect after restart.", text);
        Assert.False(DashNormalizer.ContainsDash(text));
    }

    [Fact]
    public void Default_writing_style_and_frontier_prompt_are_themselves_dash_free()
    {
        // A prompt that models em-dash punctuation teaches the model to imitate it.
        Assert.False(DashNormalizer.ContainsDash(CleanupPrompt.DefaultWritingStyle));
        Assert.False(DashNormalizer.ContainsDash(CleanupPrompt.DefaultFrontierPrompt));
        Assert.False(DashNormalizer.ContainsDash(CleanupPrompt.SingleLineWritingStyle));
    }
}

File: macos/Scribe/Tests/ScribeTests/DashNormalizerTests.swift
import XCTest

@testable import Scribe

/// The writing style asks the model not to emit em or en dashes, but that is advisory. These pin the deterministic
/// backstop that makes the house style hold whichever model answered. Ported case for case from Windows'
/// `Scribe.Core.Tests.DashNormalizerTests`; the dashes are written as escapes, since the repository bans the
/// characters themselves.
final class DashNormalizerTests: XCTestCase {
    private let em = "\u{2014}"
    private let en = "\u{2013}"

    func testLeavesTextWithoutDashesUntouched() {
        let input = "A normal sentence, with a comma; and a semicolon."
        XCTAssertEqual(DashNormalizer.normalize(input), input)
    }

    func testHyphensAndProductNamesSurvive() {
        let input = "Use GPT-5.6-Terra with Qwen3-14B on a well-structured run-on sentence."
        XCTAssertEqual(DashNormalizer.normalize(input), input)
    }

    func testSpacedEmDashBecomesAComma() {
        XCTAssertEqual(
            DashNormalizer.normalize("The pill shows Scribe is listening \(em) and where it lives."),
            "The pill shows Scribe is listening, and where it lives.")
    }

    func testUnspacedEmDashBecomesACommaAndASpace() {
        XCTAssertEqual(DashNormalizer.normalize("It works\(em)mostly."), "It works, mostly.")
    }

    func testPairedEmDashesBothBecomeCommas() {
        XCTAssertEqual(
            DashNormalizer.normalize("Say a phrase \(em) like this one \(em) and Scribe types it."),
            "Say a phrase, like this one, and Scribe types it.")
    }

    func testEnDashBetweenNumbersBecomesARangeWord() {
        XCTAssertEqual(DashNormalizer.normalize("about 1\(en)2 GB"), "about 1 to 2 GB")
        XCTAssertEqual(DashNormalizer.normalize("pages 3 \(en) 7"), "pages 3 to 7")
    }

    func testDoesNotDoubleUpExistingPunctuation() {
        XCTAssertEqual(DashNormalizer.normalize("Wait, \(em) I mean the park."), "Wait, I mean the park.")
        XCTAssertEqual(DashNormalizer.normalize("Done. \(em) Next up."), "Done. Next up.")
    }

    func testLineLeadingDashIsDroppedRatherThanTurnedIntoAComma() {
        XCTAssertEqual(DashNormalizer.normalize("Items:\n\(em) first\n\(em) second"), "Items:\nfirst\nsecond")
    }

    func testTrailingDashIsRemovedWithoutAddingPunctuation() {
        XCTAssertEqual(DashNormalizer.normalize("An unfinished thought \(em)"), "An unfinished thought")
    }

    func testDashBeforeAClosingBracketOrQuoteIsDroppedNotTurnedIntoAComma() {
        XCTAssertEqual(DashNormalizer.normalize("(an aside \(em))"), "(an aside)")
        XCTAssertEqual(DashNormalizer.normalize("he said \"fine\(em)\""), "he said \"fine\"")
        XCTAssertEqual(DashNormalizer.normalize("[note\(em)]"), "[note]")
    }

    func testDashAtTheEndOfALineDoesNotLeaveADanglingComma() {
        XCTAssertEqual(
            DashNormalizer.normalize("First thought \(em)\nSecond thought"), "First thought\nSecond thought")
        XCTAssertEqual(DashNormalizer.normalize("First\(em)\r\nSecond"), "First\r\nSecond")
    }

    func testNewlinesArePreserved() {
        XCTAssertEqual(
            DashNormalizer.normalize("First para \(em) with an aside.\n\nSecond para."),
            "First para, with an aside.\n\nSecond para.")
    }

    func testRunsOfDashesCollapseToOneReplacement() {
        XCTAssertEqual(DashNormalizer.normalize("Yes \(em)\(em) really."), "Yes, really.")
        XCTAssertEqual(DashNormalizer.normalize("Yes \(em)\(en) really."), "Yes, really.")
    }

    func testEmptyTextIsSafe() {
        XCTAssertEqual(DashNormalizer.normalize(""), "")
    }

    func testOutputNeverContainsADash() {
        let cases = [
            "a \(em) b",
            "a\(em)b",
            "\(em)leading",
            "trailing\(em)",
            "1\(en)2",
            "one \(em) two \(em) three \(em) four",
            "\(em)",
            " \(em) ",
            "(aside \(em))",
            "line \(em)\nnext",
            "quote\(em)\"",
            // A combining mark after the dash makes one Character of the two; the dash is still found.
            "a \(em)\u{301} b",
        ]

        for input in cases {
            XCTAssertFalse(DashNormalizer.containsDash(DashNormalizer.normalize(input)), input)
        }
    }

    /// The real guarantee: whatever the model answered, what the guard accepts has no dashes.
    func testSanitizedCleanupOutputIsDashFree() {
        let result = CleanupResponseGuard.sanitize(
            candidate: "The defaults are right for most people \(em) changes take effect after restart.",
            original: "the defaults are right for most people changes take effect after restart")

        XCTAssertEqual(result, .accepted("The defaults are right for most people, changes take effect after restart."))
    }

    /// A refusal stays a refusal whatever punctuation the model used: the guard judges the answer before any dash
    /// comes out of it.
    func testARefusalWithADashIsStillRejected() {
        let result = CleanupResponseGuard.sanitize(
            candidate: "I'm sorry \(em) I cannot assist with that request.",
            original: "Please schedule the meeting for tomorrow morning.")

        XCTAssertEqual(result, .rejected(.refusalLike))
    }

    /// A prompt that models dash punctuation teaches the model to imitate it.
    func testThePromptsAreThemselvesDashFree() {
        XCTAssertFalse(DashNormalizer.containsDash(CleanupPrompt.defaultWritingStyle))
        XCTAssertFalse(DashNormalizer.containsDash(CleanupPrompt.defaultFrontierPrompt))
        XCTAssertFalse(DashNormalizer.containsDash(CleanupPrompt.defaultLocalPrompt))
        XCTAssertFalse(DashNormalizer.containsDash(UsageInsight.systemPrompt))
    }
}
